sklearn/compose/_column_transformer.py

```diff
@@ -467,6 +469,8 @@ def transform(self, X):
         """
         check_is_fitted(self, 'transformers_')
 
+        X = check_array(X)
```

Comment:
A naive use of check_array will break pandas support. I think we should do something like:

```python
if not hasattr(X, 'loc'):
    X = check_array(X)
```

We don't want to do an explicit `isinstance` to avoid introducing an explicit dependency on the pandas library.

Reply:
SShouldn't such a check be done in `check_array`? Or else we should have this check almost everywhere if we want to keep the dataframe, don't we?

Reply:
Further, `check_array` should also not convert to numeric or raise on sparse in this case (it should basically do no check at all). Therefore, in this case, I would maybe not use check_array?

I think in the end we want to improve `check_array` to also handle pandas dataframes, but that is a larger endeavour than fixing this bug. 
@adrinjalali the column transformer is currently the only sklearn object where we don't convert the passed dataframe to an array, therefore this issue does not come up everywhere else

Reply:
Perhaps

```py
if not hasattr(X, '__array__'):
    X = np.asarray(X)
```
?

(or do we want dtype='object' to be safe?)

Reply:
Ah, sparse matrices apparently have no `__array__`, so an additional check should include that as well

Reply:
> Shouldn't such a check be done in check_array?

Opened a new issue about this here: https://github.com/scikit-learn/scikit-learn/issues/12148